core/src/main/java/hudson/widgets/buildsearch/BuildSearchParams.java

```diff
@@ -0,0 +1,174 @@
+
+    private static Map<String, List<BuildSearchParam>> newParamMap() {
+        Map<String, List<BuildSearchParam>> map = new LinkedHashMap<String, List<BuildSearchParam>>();
+        for (String searchTerm: BuildSearchParamProcessorFactory.getAllSearchTerms()) {
+            map.put(searchTerm, new ArrayList<BuildSearchParam>());
+        }
+        return Collections.unmodifiableMap(map);
+    }
+
+    public class BuildSearchParam {
+
+        public String searchTerm;
+        public String searchTermToken;
+
+        private String param;
+        // The start index (in this.searchString) of the search parameter (including the searchTermToken itself).
+        int startIndex;
+        // The end index (in this.searchString) of the search parameter, -1 being the end of the string.
+        int endIndex = -1;
+
+        BuildSearchParam(String searchTerm, String searchTermToken, int startIndex) {
+            this.searchTerm = searchTerm;
+            this.searchTermToken = searchTermToken;
+            this.startIndex = startIndex;
+        }
+
+        public String get() {
+            if (this.param != null) {
+                return this.param;
+            }
+
+            String param = searchString.substring(getEndOfTokenIndex(), getParamEndIndex());
+            param = param.trim();
+
+            if (parsingComplete) {
+                // We can cache the value if parsing is complete.
+                this.param = param;
+            }
+
+            return param;
+        }
+
+        private boolean isWhitespace() {
+            int paramStartIndex = getEndOfTokenIndex();
+            int paramEndIndex = getParamEndIndex();
+            for (int i = paramStartIndex; i < paramEndIndex; i++) {
+                if (!Character.isWhitespace(searchString.charAt(i))) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        private int getEndOfTokenIndex() {
+            return (startIndex + searchTermToken.length());
+        }
+
+        private int getParamEndIndex() {
+            return (endIndex != -1 ? endIndex : searchString.length());
+        }
+
+        @Override
+        public String toString() {
+            return searchTerm + ":'" + get() + "'";
+        }
+    }
+
+    public BuildSearchParams(@Nonnull String searchString) {
+        this.searchString = searchString;
+        if (searchString.trim().length() > 0) {
+            parseSearchString();
+        }
+    }
+
+    public List<BuildSearchParam> getParams(@Nonnull String searchTerm) {
+        List<BuildSearchParam> params = searchParamsSets.get(searchTerm);
+        if (params == null) {
+            throw new IllegalArgumentException(String.format("Unknown search term '%s'.", searchTerm));
+        }
+        return params;
+    }
+
+    @Override
+    public String toString() {
+        return searchParamsSets.toString();
+    }
+
+    private void parseSearchString() {
```

Comment:
I suppose the manual search implementations here could be replaced by common methods available in Apache Commons